File: frontend/src/utils/geo.js
export function haversineKm(lat1, lng1, lat2, lng2) {
  const toRad = (deg) => (deg * Math.PI) / 180;
  const R = 6371;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

export function requestUserLocation() {
  return new Promise((resolve, reject) => {
    if (!navigator?.geolocation) {
      reject(new Error('Geolocation is not supported by your browser'));
      return;
    }
    navigator.geolocation.getCurrentPosition(
      (pos) => resolve({ lat: pos.coords.latitude, lng: pos.coords.longitude }),
      (err) => {
        const messages = {
          1: 'Location permission denied. Enable location access to search nearby.',
          2: 'Unable to determine your location.',
          3: 'Location request timed out. Try again.',
        };
        reject(new Error(messages[err.code] || 'Could not get your location'));
      },
      { enableHighAccuracy: false, timeout: 12000, maximumAge: 300000 },
    );
  });
}

File: backend/services/geo.py
from __future__ import annotations

import math


def haversine_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    r = 6371.0
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlambda = math.radians(lng2 - lng1)
    a = math.sin(dphi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(dlambda / 2) ** 2
    return r * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def filter_by_distance(
    rooms: list[dict],
    lat: float,
    lng: float,
    radius_km: float = 50.0,
) -> list[dict]:
    filtered: list[dict] = []
    for room in rooms:
        loc = room.get("location") or {}
        rlat, rlng = loc.get("lat"), loc.get("lng")
        if rlat is None or rlng is None:
            continue
        distance = haversine_km(lat, lng, float(rlat), float(rlng))
        if distance <= radius_km:
            room = {**room, "distance_km": round(distance, 1)}
            filtered.append(room)
    filtered.sort(key=lambda r: r.get("distance_km", 9999))
    return filtered
